### ISSUE TITLE
mercurial regression since 16.09: does not read $out/etc/mercurial/hgrc

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#21563]
## Issue description

mercural-3.8.2 from nixpkgs-16.09 reads ```/nix/store/*-mercurial-3.8.2/etc/mercurial/hgrc```:
```
# nix-build https://github.com/NixOS/nixpkgs/archive/16.09.tar.gz -A mercurial
....
# strace -f result/bin/hg 2>&1 >/dev/null | grep hgrc
open("/nix/store/gjmm0kyfh99xsbsdswa09cc2346wd6lb-mercurial-3.8.2/etc/mercurial/hgrc.d",
O_RDONLY) = -1 ENOENT (No such file or directory)
open("/etc/mercurial/hgrc.d", O_RDONLY) = -1 ENOENT (No such file or
directory)
open("/nix/store/gjmm0kyfh99xsbsdswa09cc2346wd6lb-mercurial-3.8.2/etc/mercurial/hgrc",
O_RDONLY) = 4
open("/etc/mercurial/hgrc", O_RDONLY)   = -1 ENOENT (No such file or
directory)
open("/root/.hgrc", O_RDONLY)           = -1 ENOENT (No such file or
directory)
```

mercural-3.9.2 from nixpkgs-17.03-unstable does not:
```
# nix-build https://github.com/NixOS/nixpkgs/archive/master.tar.gz -A mercurial
....
# strace -f result/bin/hg 2>&1 >/dev/null | grep hgrc
open("etc/mercurial/hgrc.d", O_RDONLY)  = -1 ENOENT (No such file or
directory)
open("/etc/mercurial/hgrc.d", O_RDONLY) = -1 ENOENT (No such file or
directory)
open("etc/mercurial/hgrc", O_RDONLY)    = -1 ENOENT (No such file or
directory)
open("/etc/mercurial/hgrc", O_RDONLY)   = -1 ENOENT (No such file or
directory)
open("/root/.hgrc", O_RDONLY)           = -1 ENOENT (No such file or directory)
```
thus, no hgrc found at all

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
NixOS 17.03 unstable
Linux Mint 18

* Nix version: (run `nix-env --version`)
1.11.4

* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
"17.03pre98014.5f1ebf8"

**Comment by volth** [comment]
@FRidh Any reason to use ```basename``` [here](https://github.com/NixOS/nixpkgs/blame/31dc66a55ae2a062b7f025879808afe30631fff8/pkgs/development/interpreters/python/wrap-python.nix#L38) instead of full path in /nix/store ?

Various applications patch the generated wrapper code to full path:  [1](
https://github.com/NixOS/nixpkgs/blob/31dc66a55ae2a062b7f025879808afe30631fff8/pkgs/applications/office/zim/default.nix#L35) [2](https://github.com/NixOS/nixpkgs/blob/31dc66a55ae2a062b7f025879808afe30631fff8/pkgs/applications/misc/calibre/default.nix#L80) 

This issue requires the same fix for mercurial.

Wouldn't it be better to fix ```wrap-python.nix``` ?

**Comment by FRidh** [comment]
@volth you mean, which set `sys.argv[0]` to the name of the script, and not to the path?

If I am correct on *nix-based systems `sys.argv[0]` returns the name of the script, whereas on Windows it returns the full path.

Would it be a problem if we switch to the full path as well? I don't know - it's definitely worth giving a try - but I wouldn't be surprised if some programs break because of it. I mean, these two programs already show that they're doing something they shouldn't; that is, rely on `sys.argv[0]` returning a full path.

**Comment by FRidh** [comment]
The [Python docs](https://docs.python.org/3.5/library/sys.html?highlight=sys.argv#sys.argv) state the following 

>  argv[0] is the script name (it is operating system dependent whether this is a full pathname or not)

In my opinion it is really bad programs depend on it and it should be considered a bug. 

In any case, it is interesting to see that the full path is apparently used on a Linux distribution as well. Therefore, I think we should investigate what other big distributions do, and give the change a try on Nix as well. Furthermore it might be worth checking the Python mailing lists on this topic.

**Comment by volth** [comment]
@FRidh offtopic: is [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/mercurial/default.nix#L41-L44) necessary? After you [changed](https://github.com/NixOS/nixpkgs/commit/a01d7d131edb31513123eaa0def8fb8abca60430) ```stdenv.mkDerivation``` to ```python2Packages.mkPythonDerivation```, ```hg``` is wrapped twice. There are ```hg```, ```.hg-wrapper``` and ```..hg-wrapper-wrapper```

**Comment by FRidh** [comment]
Python itself also provides the full path on Nix, so therefore I think we should fix `wrap-python.nix`.

> @FRidh offtopic: is this necessary? After you changed stdenv.mkDerivation to python2Packages.mkPythonDerivation, hg is wrapped twice. There are hg, .hg-wrapper and ..hg-wrapper-wrapper

This happens quite often with Nix that we have wrappers applied several times (we really should use Nix for composing the wrapper and call `wrapProgram` only once). Generally this doesn't cause any problems though.

**Comment by vcunat** [comment]
It would be easiest if `makeWrapper` recognized its own products and composed the wrappers automatically.

**Comment by volth** [comment]
@vcunat I have the same result with ```bash``` as I had with ```fish```. 
Moreover, if the program is not run like ```python a.py```, but has first line ```#!/usr/bin/python``` and executable bit ```sys.path[0]``` is always path. At least ```./a.py```, but never just a file name.

**Comment by vcunat** [comment]
I've used `realpath "$0"` in some scripts to locate themselves and so far I haven't found a case where it wouldn't work (even if it's not guaranteed by standards).

**Comment by volth** [comment]
> Generally this doesn't cause any problems though.

It does not cause problem but it devalues all the trickery with ```sys.argv[0]``` (which goal is to [Sets argv[0] to the original application's name; otherwise it would be .foo-wrapped.](https://github.com/NixOS/nixpkgs/blame/31dc66a55ae2a062b7f025879808afe30631fff8/pkgs/development/interpreters/python/wrap-python.nix#L31)) in case of mercurial there is ```sys.argv[0] = '.hg-wrapped'```

**Comment by FRidh** [comment]
See https://github.com/NixOS/nixpkgs/pull/21568 where I updated `wrap-python.nix` to include the full path. 
It won't solve the 
```
in case of mercurial there is sys.argv[0] = '.hg-wrapped'
```
issue but I don't see any issue around that. Using [`buildEnv`](https://github.com/NixOS/nixpkgs/pull/16672) won't solve that entirely either.

**Comment by FRidh** [comment]
@volth with https://github.com/NixOS/nixpkgs/pull/21568
```
$ strace -f result/bin/hg 2>&1 >/dev/null | grep hgrc
open("/nix/store/9ifg86qrj9f35sypdvjaisngf2lpzpds-python2.7-mercurial-3.9.2/etc/mercurial/hgrc.d", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/etc/mercurial/hgrc.d", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/nix/store/9ifg86qrj9f35sypdvjaisngf2lpzpds-python2.7-mercurial-3.9.2/etc/mercurial/hgrc", O_RDONLY) = 3
open("/etc/mercurial/hgrc", O_RDONLY)   = -1 ENOENT (No such file or directory)
open("/home/freddy/.hgrc", O_RDONLY)    = -1 ENOENT (No such file or directory)
```

**Comment by FRidh** [comment]
Closing since this is fixed as shown in https://github.com/NixOS/nixpkgs/issues/21563#issuecomment-269919402

**Comment by volth** [comment]
The change is not in ```master``` branch yet.
If ```wrap-python.nix``` change is not approved, then I have to add a zim-like patch for mercurial.

**Comment by FRidh** [comment]
The fixes are in staging and should end up in master in a couple of days.

**Comment by FRidh** [comment]
> As far as I can tell all Python packages still function (https://headcounter.org/hydra/eval/344688) with this change.

> I've pushed d68aba4 and a zim fix 21f6cb7.
